### PR TITLE
KAFKA-17447: Changed fetch queue processing to reduce the no. of locking and unlocking activity

### PR DESCRIPTION
**Pull request by adixitconfluent**
### About
For the share groups fetch request processing, we have an recursive approach of dealing with individual fetch requests. While  it works fine with less no. of records (< 1,000,000) and lesser sharing (< 5 share consumers), it seems that some requests are getting stuck when we increase the load and try to increase the throughput. I've replaced this approach by removing the unlocking and locking of fetch queue in between entries. This had reduced the complexity and also removes the reliability issue on increasing the load.

### Testing
The code has been tested with the help of unit tests.

### CODE REVIEW

File: core/src/main/java/kafka/server/share/SharePartitionManager.java
Comment (AndrewJSchofield) [inline core/src/main/java/kafka/server/share/SharePartitionManager.java]: nit: `releasePartitionLocks` would be a better name.

File: core/src/main/java/kafka/server/share/SharePartitionManager.java
Comment (AndrewJSchofield) [inline core/src/main/java/kafka/server/share/SharePartitionManager.java:529]: I believe that `processFetchQueue()` always releases the `processFetchQueueLock` is the final step when it returns. I would prefer to see the lock released in a try-finally in this `maybeProcessFetchQueue` method to make the code more resilient to change. Having the lock and unlock in separate methods seems unwise, and I believe the code structure makes this quite simple.

**Review by AndrewJSchofield — CHANGES_REQUESTED** [review]
Thanks for the PR. A couple of minor comments. I prefer the new structure.

**Comment by apoorvmittal10** [comment]
I will wait for the PR to be merged as we are moving to purgatory based fetch. @adixitconfluent please let me know if we should irrespectively review and merge this PR.

**Comment by adixitconfluent** [comment]
Hi @apoorvmittal10 , this PR might not be needed since we're moving to a purgatory based implementation. I'll put it in draft mode for now until we have the purgatory PR merged.

**Comment by AndrewJSchofield** [comment]
@adixitconfluent I note that the purgatory code remains recursive so we might choose to take this improvement too.

**Comment by adixitconfluent** [comment]
hi @AndrewJSchofield, yes, but as advised, we'll get rid of the fetch queue altogether and the purgatory is going to receive the share fetch requests directly. Hence, we won't be needing this PR now.

**Pull request closed without merging**
